src/zenml/utils/metadata_utils.py

```diff
@@ -0,0 +1,330 @@
+#  Copyright (c) ZenML GmbH 2024. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Utility functions to handle metadata for ZenML entities."""
+
+from typing import Dict, Optional, Set, Union, overload
+from uuid import UUID
+
+from zenml.client import Client
+from zenml.enums import MetadataResourceTypes
+from zenml.logger import get_logger
+from zenml.metadata.metadata_types import MetadataType
+from zenml.steps.step_context import get_step_context
+
+logger = get_logger(__name__)
+
+
+@overload
+def log_metadata(metadata: Dict[str, MetadataType]) -> None: ...
```

Comment:
I'm curious what all those overloads give us here. Params are optional anyway...

Reply:
You are right, in the original implementation, all the parameters are optional but the overloads here give us an advantage in two significant ways. 

- It helps us with `mypy` linting to see if the function is used in anyway that it shouldn't be.
- It helps the developer as well. If I write `log_metadata(metadata={}, artifact_name=..., step_name=...)`, my IDE warns me that this is not an allowed usage of the function.